plugins/outputs/dynatrace/dynatrace_test.go

```diff
@@ -0,0 +1,214 @@
+package dynatrace
+
+import (
+	"encoding/json"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/testutil"
+	"github.com/stretchr/testify/require"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestNilMetrics(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(`{"linesOk":10,"linesInvalid":0,"error":null}`)
+	}))
+	defer ts.Close()
+
+	d := &Dynatrace{}
+	d.EnvironmentURL = ts.URL
+	d.EnvironmentAPIToken = "123"
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.NoError(t, err)
+
+	err = d.Write(nil)
+	require.NoError(t, err)
+}
+
+func TestEmptyMetricsSlice(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(`{"linesOk":10,"linesInvalid":0,"error":null}`)
+	}))
+	defer ts.Close()
+
+	d := &Dynatrace{}
+	d.EnvironmentURL = ts.URL
+	d.EnvironmentAPIToken = "123"
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.NoError(t, err)
+	empty := []telegraf.Metric{}
+	err = d.Write(empty)
+	require.NoError(t, err)
+}
+
+func TestMockURL(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(`{"linesOk":10,"linesInvalid":0,"error":null}`)
+	}))
+	defer ts.Close()
+
+	d := &Dynatrace{}
+	d.EnvironmentURL = ts.URL
+	d.EnvironmentAPIToken = "123"
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.NoError(t, err)
+
+	err = d.Write(testutil.MockMetrics())
+	require.NoError(t, err)
+}
+
+func TestMissingURL(t *testing.T) {
+	d := &Dynatrace{}
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.Equal(t, oneAgentMetricsUrl, d.EnvironmentURL)
+	require.NoError(t, err)
+}
+
+func TestMissingAPITokenMissingURL(t *testing.T) {
+	d := &Dynatrace{}
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.Equal(t, oneAgentMetricsUrl, d.EnvironmentURL)
+	require.NoError(t, err)
+}
+
+func TestMissingAPIToken(t *testing.T) {
+	d := &Dynatrace{}
+	d.EnvironmentURL = "test"
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.Error(t, err)
+}
+
+func TestSendMetric(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(`{"linesOk":10,"linesInvalid":0,"error":null}`)
```

Comment:
would be good to test the actual encoded data here to make sure it's formatting it properly. I'd like to see that for most of the outputs. Testing it connects isn't really enough.

Reply:
I added encoding tests within all outputs, according to your request.